From: <Joerg-Cyril.Hoehle@t-systems.com>
Subject: ntdll: Use the monotonic time counter also for timer queues.
Message-Id: <8E4C156DA5797D418DBFADFD8CE655A41FA3C5BDBA@HE113481.emea1.cds.t-internal.com>
Date: Thu, 7 Feb 2013 15:00:27 +0100

Hi,

Wine should not base periodic timers nor audio on a TOD clock that
is subject to DST and other time adjustments like NTP.

>now.QuadPart * 1000 / freq.QuadPart;
I don't like this X * numerator / denominator
business that transcends all timers that much.
In effect, it means that we loose a significant amount of bits
(depending on how large num. and den. are)
from the presumably soo large 64bit values.

That pattern is present in several places in Wine.

An opportunity for optimization would be
if (denominator = 10000000) /* short-circuit the known Wine value */
  return now / 10000
 else return now * num / den

I've tested this transition using the tests in ntdll, kernel, mmdevapi and winmm.

Regards,
 Jörg Höhle

From ee313aaf51ce5cd8ee8706b68f33d4c8218c1610 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?J=C3=B6rg=20H=C3=B6hle?= <hoehle@users.sourceforge.net>
Date: Tue, 5 Feb 2013 18:01:04 +0100
Subject: ntdll: Use the monotonic time counter also for timer queues.

---
 dlls/ntdll/threadpool.c |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/dlls/ntdll/threadpool.c b/dlls/ntdll/threadpool.c
index 5c1a571..e33a025 100644
--- a/dlls/ntdll/threadpool.c
+++ b/dlls/ntdll/threadpool.c
@@ -605,9 +605,9 @@ static DWORD WINAPI timer_callback_wrapper(LPVOID p)
 
 static inline ULONGLONG queue_current_time(void)
 {
-    LARGE_INTEGER now;
-    NtQuerySystemTime(&now);
-    return now.QuadPart / 10000;
+    LARGE_INTEGER now, freq;
+    NtQueryPerformanceCounter(&now, &freq);
+    return now.QuadPart * 1000 / freq.QuadPart;
 }
 
 static void queue_add_timer(struct queue_timer *t, ULONGLONG time,
-- 
1.5.6.3


